### PR TITLE
Add new illegal_subcategory field in abuse reports

### PR DESCRIPTION
**Pull request by willdurand**
Fixes https://github.com/mozilla/addons/issues/14875

~~:warning: depends on https://github.com/mozilla/addons-server/pull/22388~~

### CODE REVIEW

**Comment by eviljeff** [comment]
can haz rebase after #22388 ?

File: src/olympia/abuse/serializers.py [hunk 1]
```diff
@@ -93,6 +103,38 @@ def validate(self, data):
             elif data.get('illegal_category') is None:
                 msg = serializers.Field.default_error_messages['null']
                 raise serializers.ValidationError({'illegal_category': [msg]})
+        elif data.get('illegal_category') is not None:
+            msg = (
+                'This value may be omitted or set to "null" when the reason is not '
+                '"illegal".'
+            )
+            raise serializers.ValidationError({'illegal_category': [msg]})
```
Comment (willdurand) [inline src/olympia/abuse/serializers.py:111]: That's something I forgot to check in the previous PR..

File: src/olympia/abuse/models.py [hunk 2]
```diff
@@ -638,6 +639,13 @@ class AbuseReport(ModelBase):
         null=True,
         help_text='Type of illegal content',
     )
+    illegal_subcategory = models.PositiveSmallIntegerField(
+        default=None,
+        choices=ILLEGAL_SUBCATEGORIES.choices,
+        blank=True,
+        null=True,
+        help_text='Specific violation',
```
Comment (eviljeff) [inline src/olympia/abuse/models.py:647]:
```suggestion
        help_text='Specific violation of illegal content',
```
or some better wording.  Just to make it clear it's connected to the `illegal_category`

File: src/olympia/abuse/serializers.py [hunk 3]
```diff
@@ -93,6 +103,38 @@ def validate(self, data):
             elif data.get('illegal_category') is None:
                 msg = serializers.Field.default_error_messages['null']
                 raise serializers.ValidationError({'illegal_category': [msg]})
+        elif data.get('illegal_category') is not None:
+            msg = (
+                'This value may be omitted or set to "null" when the reason is not '
+                '"illegal".'
```
Comment (eviljeff) [inline src/olympia/abuse/serializers.py:109]:
```suggestion
                'This value must be omitted or set to "null" when the `reason` is not '
                '"illegal".'
```

File: src/olympia/abuse/serializers.py [hunk 4]
```diff
@@ -93,6 +103,38 @@ def validate(self, data):
...
+            elif subcategory not in valid_subcategories:
+                msg = (
+                    'This value cannot be used in combination with the '
+                    'supplied illegal category.'
+                )
+                raise serializers.ValidationError({'illegal_subcategory': [msg]})
+        elif data.get('illegal_subcategory') is not None:
+            msg = (
+                'This value may be omitted or set to "null" when the reason is not '
+                '"illegal".'
```
Comment (eviljeff) [inline src/olympia/abuse/serializers.py:135]:
```suggestion
                'This value must be omitted or set to "null" when the `reason` is not '
                '"illegal".'
```

File: src/olympia/abuse/serializers.py [hunk 5]
```diff
@@ -93,6 +103,38 @@ def validate(self, data):
...
+            if 'illegal_subcategory' not in data:
+                msg = serializers.Field.default_error_messages['required']
+                raise serializers.ValidationError({'illegal_subcategory': [msg]})
+            elif subcategory is None:
+                msg = serializers.Field.default_error_messages['null']
+                raise serializers.ValidationError({'illegal_subcategory': [msg]})
+            elif subcategory not in valid_subcategories:
+                msg = (
+                    'This value cannot be used in combination with the '
+                    'supplied illegal category.'
```
Comment (eviljeff) [inline src/olympia/abuse/serializers.py:129]:
```suggestion
                    'This value cannot be used in combination with the '
                    'supplied `illegal_category`.'
```

File: src/olympia/abuse/tests/test_views.py
Comment (eviljeff) [inline src/olympia/abuse/tests/test_views.py:1]: reminder to update these tests if any of the suggestions in serializers.py are accepted

**Review by eviljeff — APPROVED** [review]

**Pull request merged**